### PR TITLE
tools/mpremote: Add `mpremote mip install` to install packages.

### PR DESCRIPTION
**Pull request by jimmo**
This supports the same package sources as the new `mip` tool. (https://github.com/micropython/micropython-lib/pull/542)
 - micropython-lib (by name)
 - http(s) & github packages with json description
 - directly downloading a .py/.mpy file

The version is specified with an optional `@version` on the end of the package name.

Also adds support for optional --args to multi-arg commands like "edit" and the filesystem commands (and now "install"). This allows `--mpy`/`--no-mpy` and `--target` to be specified to the `install` command.

_This work was funded through GitHub Sponsors._

### CODE REVIEW

**Comment by jimmo** [comment]
This depends on #9463 for the reference ` :ref:``packages`` ` to make sense.

**Comment by jimmo** [comment]
Updated to make `install` a subcommand of `mip`. We will likely add more subcommands of `mip` in the future (e.g. `search`, `download`, etc).

File: tools/mpremote/mpremote/main.py [hunk 1]
```diff
@@ -213,421 +400,102 @@ def usage_error(cmd, exp_args, msg):
...
-                pyb = do_connect(["auto"])
-            if pyb.in_raw_repl:
-                pyb.exit_raw_repl()
-            do_repl(pyb, args)
+            # Get any leftover unprocessed args.
+            remaining_args = args.next_command + extra_args
+
+        # If no commands were "actions" then implicitly finish with the REPL.
+        if state.run_repl_on_completion():
+            do_repl(state)
```
Comment (dpgeorge) [inline tools/mpremote/mpremote/main.py:494]: this call fails because it's defined as `do_repl(state, args)`

**Comment by dpgeorge** [comment]
A few things I noticed while testing that were "unexpected":
- no progress display when mip is copying files
- mip will redownload and recopy files that already exist on the target

Maybe these are things that will be improved in a future PR?

File: tools/mpremote/mpremote/main.py (hunk 1 above)
Comment (jimmo) [reply tools/mpremote/mpremote/main.py]: Accidental cleanup in one of the later updates. Fixed.

**Comment by jimmo** [comment]
> * no progress display when mip is copying files

Done. There's now a progress for downloading then another one for copying.

> * mip will redownload and recopy files that already exist on the target

I'm not sure this is important for the PC version of `mip`. But it would be easy to add in a future PR.

**Comment by dpgeorge** [comment]
A great feature, thank you!

**Comment by peterhinch** [comment]
> mip will redownload and recopy files that already exist on the target

In the case where the file(s) have been updated this would be welcome behaviour.

**Comment by jimmo** [comment]
> > mip will redownload and recopy files that already exist on the target
> 
> In the case where the file(s) have been updated this would be welcome behaviour.

@peterhinch 

Yes, definitely.

What Damien is referring to here though is that `mip` when running on the device is "smarter" and will only re-download if the file on the device is different to the one in the index. (The index sends a single manifest with the required file hashes for the package and all dependencies, and it computes the hashes and decides what to fetch -- which is to say, if you `mip.install("foo")` and only a single file in a dependency changed, then there are only two total HTTP requests: the initial manifest, and that one file)

When running in `mpremote`, mip always just unconditionally downloads everything, and never computes and checks the local on-device hashes. We can definitely add the smarter behavior though if it's useful.

( A note on the list of required hashes for anyone who is curious: the list of hashes is computed at the time the package is published and includes all dependencies. i.e. version x.y.z "freezes" all it's dependencies too at that point in time, so `mpremote mip install package@version` (or `mip.install(package, version=)`) will always give you the exact same files, even if a dependency gets updated without bumping its version number. )

**Pull request merged**